Media elements: Do not use getNonEmptyURLAttribute() in media source selection

Ahmad-S792 opened the pull request:
#### dccfae42bb29bd4bdef052e469f604a9387241c0
<pre>
Media elements: Do not use getNonEmptyURLAttribute() in media source selection
<a href="https://bugs.webkit.org/show_bug.cgi?id=276509">https://bugs.webkit.org/show_bug.cgi?id=276509</a>
<a href="https://rdar.apple.com/132042925">rdar://132042925</a>

Reviewed by Eric Carlson.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Merge: <a href="https://chromium.googlesource.com/chromium/src.git/+/0f1b99def010815efd548304136aa69441161c91">https://chromium.googlesource.com/chromium/src.git/+/0f1b99def010815efd548304136aa69441161c91</a>

The specification [1] says:

&quot; If the src attribute&apos;s value is the empty string, then end the synchronous section...&quot; and onward.

This specification doesn&apos;t ask to strip whitespaces in &apos;src&apos; attribute&apos;s value
though getNonEmptyURLAttribute(srcAttr) does.

[1] <a href="https://html.spec.whatwg.org/multipage/media.html#concept-media-load-algorithm">https://html.spec.whatwg.org/multipage/media.html#concept-media-load-algorithm</a>

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::selectMediaResource):
(WebCore::HTMLMediaElement::selectNextSourceChild):
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/location-of-the-media-resource/currentSrc-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/location-of-the-media-resource/currentSrc-expected.txt: Deleted.

Canonical link: <a href="https://commits.webkit.org/289432@main">https://commits.webkit.org/289432@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/388684dbd3db3a94280b8bf72dfe1fdf1df7283f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86902 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6409 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41246 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91750 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37636 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/88951 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6677 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14473 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67180 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24949 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89905 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5087 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78655 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47498 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4872 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33021 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36752 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75374 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33900 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93643 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14055 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10210 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75978 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14256 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74503 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75175 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18493 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19495 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17914 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/6891 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14078 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19341 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13816 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17261 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15601 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -1700,13 +1700,14 @@ void HTMLMediaElement::selectMediaResource()
                 return;
             }
 
+            URL absoluteURL = document().completeURL(srcValue);
```
```suggestion
            auto absoluteURL = document().completeURL(srcValue);
```

Ahmad-S792 replied:
Done.

annevk commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -5576,12 +5577,19 @@ URL HTMLMediaElement::selectNextSourceChild(ContentType* contentType, String* ke
+        // from parsing the URL specified by candidate's src attribute's value
+        // relative to the candidate's node document when the src attribute was
+        // last changed.
+        auto mediaURL = source->document().completeURL(srcValue);
```
Shouldn't this move lower? Why would we want to run this code for the empty string value?

Ahmad-S792 replied:
It gives compiler error when I move it below `goto CheckAgain`.

Ahmad-S792 replied:
```
/Users/ahmadsaleem/OpenSource/Source/WebCore/html/HTMLMediaElement.cpp:5596:17: error: 
      cannot jump from this goto statement to its label
 5596 |                 goto CheckAgain;
      |                 ^
/Users/ahmadsaleem/OpenSource/Source/WebCore/html/HTMLMediaElement.cpp:5603:14: note: 
      jump bypasses variable initialization
 5603 |         auto mediaURL = source->document().completeURL(srcValue);
      |              ^
/Users/ahmadsaleem/OpenSource/Source/WebCore/html/HTMLMediaElement.cpp:5588:13: error: 
      cannot jump from this goto statement to its label
 5588 |             goto CheckAgain;
      |             ^
/Users/ahmadsaleem/OpenSource/Source/WebCore/html/HTMLMediaElement.cpp:5603:14: note: 
      jump bypasses variable initialization
 5603 |         auto mediaURL = source->document().completeURL(srcValue);
      |              ^
2 errors generated.
```

annevk replied:
I see. I guess you could initialize the variable first using `URL mediaURL;`, but only call the function later, but `goto` is above my pay grade so someone else will have to review this.

Ahmad-S792 replied:
@annevk - updated.

Ahmad-S792 committed:
Media elements: Do not use getNonEmptyURLAttribute() in media source selection
https://bugs.webkit.org/show_bug.cgi?id=276509
rdar://132042925

Reviewed by Eric Carlson.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Merge: https://chromium.googlesource.com/chromium/src.git/+/0f1b99def010815efd548304136aa69441161c91

The specification [1] says:

" If the src attribute's value is the empty string, then end the synchronous section..." and onward.

This specification doesn't ask to strip whitespaces in 'src' attribute's value
though getNonEmptyURLAttribute(srcAttr) does.

[1] https://html.spec.whatwg.org/multipage/media.html#concept-media-load-algorithm

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::selectMediaResource):
(WebCore::HTMLMediaElement::selectNextSourceChild):
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/location-of-the-media-resource/currentSrc-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/location-of-the-media-resource/currentSrc-expected.txt: Deleted.

Canonical link: https://commits.webkit.org/289432@main
